server/packet/processor.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package packet
+
+import (
+	"sync"
+
+	"github.com/m3db/m3aggregator/aggregator"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/sync"
+
+	"github.com/uber-go/tally"
+)
+
+type processorMetrics struct {
+	drainErrors   tally.Counter
+	processPacket instrument.MethodMetrics
+}
+
+func newProcessorMetrics(scope tally.Scope, samplingRate float64) processorMetrics {
+	return processorMetrics{
+		drainErrors:   scope.Counter("drain-errors"),
+		processPacket: instrument.NewMethodMetrics(scope, "process", samplingRate),
+	}
+}
+
+// Processor is a packet processor
+type Processor struct {
+	queue      *Queue
+	aggregator aggregator.Aggregator
+	workers    xsync.WorkerPool
+	wgWorkers  sync.WaitGroup
+	nowFn      clock.NowFn
+	log        xlog.Logger
+	metrics    processorMetrics
+}
+
+// NewProcessor creates a new processor
+func NewProcessor(
+	queue *Queue,
+	aggregator aggregator.Aggregator,
+	numWorkers int,
+	clockOpts clock.Options,
+	instrumentOpts instrument.Options,
+) *Processor {
+	scope := instrumentOpts.MetricsScope().SubScope("processor")
+	samplingRate := instrumentOpts.MetricsSamplingRate()
+	p := &Processor{
+		queue:      queue,
+		aggregator: aggregator,
+		nowFn:      clockOpts.NowFn(),
+		log:        instrumentOpts.Logger(),
+		metrics:    newProcessorMetrics(scope, samplingRate),
+	}
+
+	// Start the workers to drain the queue
+	p.wgWorkers.Add(numWorkers)
+	p.workers = xsync.NewWorkerPool(numWorkers)
+	p.workers.Init()
+	for i := 0; i < numWorkers; i++ {
+		p.workers.Go(p.drain)
+	}
+
+	return p
+}
+
+// Close closes the processor. It's safe to call close more than once.
+// All but the first call are no-ops.
+func (p *Processor) Close() {
```

Comment:
I feel it's more idiomatic to call it `Close` since it means the processor is closing itself and once closed, will be unavailable for any further processing, so I prefer to keep it unless you feel strongly.